axiom/datasets.go

```diff
@@ -379,70 +385,109 @@ func (s *DatasetsService) IngestEvents(ctx context.Context, id string, events []
 // Restrictions for field names (JSON object keys) can be reviewed here:
 // https://www.axiom.co/docs/usage/field-restrictions.
 //
-// As this method keeps a connection open until the channel is closed, it is not
-// advised to use this method for long-running ingestions.
+// Events are ingested in batches. A batch is either 1024 events for unbuffered
+// channels or the capacity of the channel for buffered channels. A batch is
+// sent to the server as soon as it is full or after one second or when the
+// channel is closed.
 func (s *DatasetsService) IngestChannel(ctx context.Context, id string, events <-chan Event, options ...ingest.Option) (*ingest.Status, error) {
+	const (
+		defaultBatchSize    = 1024
+		defaultSendInterval = time.Second
+	)
+
 	ctx, span := s.client.trace(ctx, "Datasets.IngestChannel", trace.WithAttributes(
 		attribute.String("axiom.dataset_id", id),
 		attribute.Int("axiom.channel.capacity", cap(events)),
 	))
 	defer span.End()
 
-	// Apply supplied options.
-	var opts ingest.Options
-	for _, option := range options {
-		option(&opts)
-	}
+	var (
+		errCh        = make(chan error)
+		ingestStatus ingest.Status
+	)
+	go func() {
```

Comment:
this is a very big and complex go routine - should be broken down into some smaller chunks with clearer scoping

Reply:
I refactored the whole method and this shouldn't be a problem anymore. Please re-review.

I noticed that I don't need a go-routine as the method is blocking, anyways. The idea with the go-routine came from another design I had in mind where a go-routine would drain the producer channel into a batch and the parent go-routine would swap the full batch with an empty one so it can ingest without blocking the producer channel being read.